Updating README.adoc and CONTRIBUTING.adoc [DO NOT READ] (#4928)
input-output-hk/plutus: CONTRIBUTING.adoc

> :toc: left

Author: I've taken the liberty to shrink the README, and moved all development related stuff to CONTRIBUTING
Make a Development section and moved it right below the Introduction.
My line of reasoning was: If I want to contribute, I do want a short Introduction to the project, and then I just want to  know: "How to I build this?"

:reproducible:

> == Setting up and working with our development tools

Author: This file is best reviewed rendered -> https://github.com/input-output-hk/plutus/blob/0323c8bb2919b0b7b63d9601586177380fd37bec/CONTRIBUTING.adoc

Author: The section `How to setup you development environment` has been largely re-written.
All other sections are mostly unchanged (except for the occasional amend `nix-shell` -> `nix develop`)

Author: So everything below `How to update our Haskell dependencies` is mostly unchanged

> == How to setup you development environment 

Reviewer: ```suggestion
== How to setup a development environment 
```

=== How to get a shell environment with tools

To enter the development shell run the following command in the root directory: 

`nix develop`

If you receive this message:

`error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override`

Then try this:

`nix develop --extra-experimental-features 'nix-command flakes'`

You may now choose to edit your `/etc/nix/nix.conf` (global) or `~/.config/nix/nix.conf` (you must be a https://nixos.org/nix/manual/#ssec-multi-user[trusted user] to do this) by appending the following line: 

`extra-experimental-features = nix-command flakes`

If you do so you will be able to enter the development shell by running `nix develop`.

If you are a VSCode user, you may want to start your development session by running:

`nix develop --command code`

If this is your first time running `nix develop`, you may be asked the following:
```
do you want to allow configuration setting 'allow-import-from-derivation' to be set to 'true' (y/N)? y
do you want to permanently mark this value as trusted (y/N)? y
```
Please type `y` to this and all subsequent promopts and make life easy for yourself.

> IMPORTANT: *It is particulary important to accept the `extra-subsitures` settings, which will grant access to our Nix binary cache.*

Reviewer: ```suggestion
IMPORTANT: *It is particularly important to accept the `extra-substituters` and `extra-trusted-public-keys` settings, which will grant access to our Nix binary cache.*
```

At this point and if this is your first time running `nix develop`, Nix will download, build and install a copious amout of dependencies.

It is not uncommon for this process to take a couple of hours and to write tens of GBs to the `/nix/store`.

*Rest assured that this only happens the very first time you run `nix develop`.*

However you should pay attention to the output of `nix develop`: if you notice that you are *building* GHC, it is likely that your 
caches have not been configured correctly.

> Accepting the configuration settings as outlined above should be suffienct to avoid this, however if your caches are still broken, you may try to append the following lines to `/etc/nix/nix.conf` (global) and/or `~/.config/nix/nix.conf` (user):

Reviewer: ```suggestion
Accepting the configuration settings as outlined above should be sufficient to avoid this, however if your caches are still broken, you may try to append the following lines to `/etc/nix/nix.conf` (global) and/or `~/.config/nix/nix.conf` (user):
```

----
substituters = https://hydra.iohk.io https://iohk.cachix.org https://cache.nixos.org/
trusted-public-keys = hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ= iohk.cachix.org-1:DpRUyj7h7V830dp/i6Nti+NEO2/nhblbov/8MW7Rqoo= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
extra-experimental-features = nix-command flakes
----

On NixOS, set the following NixOS options:
----
nix = {
  binaryCaches          = [ "https://hydra.iohk.io" "https://iohk.cachix.org" ];
  binaryCachePublicKeys = [ "hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ=" "iohk.cachix.org-1:DpRUyj7h7V830dp/i6Nti+NEO2/nhblbov/8MW7Rqoo=" ];
};
----

You will need to restart the nix-daemon on non-NixOS for the changes to take effect, or just restart your machine.

=== What to do once inside the `nix develop` shell

Your prompt will change to `[plutus-shell]` and you will be presented with a menu of available commands.

Please read that menu carefully.

> Among the included tools are: `cabal`, `stylish-haskell`, `Agda`, `R`.

Reviewer: redundant?

[[build-with-nix]]
=== How to build the project's artifacts with Nix

Haskell components are provisioned via link:https://github.com/input-output-hk/haskell.nix[Haskell.nix]

In general you can run `nix build .#SYSTEM.plutus.library.plutus-project.hsPkgs.PACKAGE.components.COMPONENT` 

For example `nix build .#x86_64-linux.plutus.library.plutus-project.hsPkgs.plutus-core.components.library`

For full documentation about the Nix code see the link:nix/README.md[Nix `README`].

There you will see how to build all other artifacts, which are mostly related to documentation.

[[develop-without-nix]]
=== How to develop without Nix

You can build some of the Haskell packages without Nix, but this is not recommended and we don't guarantee that these prerequisites are sufficient.

If you use Nix, these tools are provided for you via `nix develop`, and you do *not* need to install them yourself.

* If you want to build our Haskell packages with https://www.haskell.org/cabal/[`cabal`], then install it.
* If you want to build our Agda code, then install https://github.com/agda/agda[Agda] and the https://github.com/agda/agda-stdlib[standard library].

[WARNING]
====

> You can also use `cabal` outside the `nix develop` environment to build the project.

Reviewer: This now reads a bit redundantly with the text above.

bug workarounds.
====

=== How to add a new Haskell package

You need to do a few things when adding a new package, in the following order:

- Add the cabal file for the new package.
- Add the package to link:cabal.project[`cabal.project`].

> - Check that you can build the package with nix as well (see xref:build-with-nix[How to build with Nix])

Reviewer: ```suggestion
- Check that you can build the package with nix as well (see xref:build-with-nix[How to build with Nix]). Or, you can wait for CI to check this for you, it should normally just work.
```
